stdlib/multiprocessing/connection.pyi

```diff
@@ -22,54 +25,54 @@ class _ConnectionBase:
+        self, exc_type: type[BaseException] | None, exc_value: BaseException | None, exc_tb: TracebackType | None
     ) -> None: ...
-    def __del__(self) -> None: ...
 
-class Connection(_ConnectionBase): ...
+class Connection(_ConnectionBase[_T1, _T2]): ...
 
 if sys.platform == "win32":
-    class PipeConnection(_ConnectionBase): ...
+    class PipeConnection(_ConnectionBase[_T1, _T2]): ...
 
 class Listener:
     def __init__(
         self, address: _Address | None = None, family: str | None = None, backlog: int = 1, authkey: bytes | None = None
     ) -> None: ...
-    def accept(self) -> Connection: ...
+    def accept(self) -> Connection[Incomplete, Incomplete]: ...
     def close(self) -> None: ...
     @property
     def address(self) -> _Address: ...
     @property
     def last_accepted(self) -> _Address | None: ...
     def __enter__(self) -> Self: ...
     def __exit__(
-        self, exc_type: type[BaseException] | None, exc_value: BaseException | None, exc_tb: types.TracebackType | None
+        self, exc_type: type[BaseException] | None, exc_value: BaseException | None, exc_tb: TracebackType | None
     ) -> None: ...
 
+# Any: send and recv methods unused
 if sys.version_info >= (3, 12):
-    def deliver_challenge(connection: Connection, authkey: bytes, digest_name: str = "sha256") -> None: ...
+    def deliver_challenge(connection: Connection[Any, Any], authkey: bytes, digest_name: str = "sha256") -> None: ...
 
 else:
-    def deliver_challenge(connection: Connection, authkey: bytes) -> None: ...
+    def deliver_challenge(connection: Connection[Any, Any], authkey: bytes) -> None: ...
 
-def answer_challenge(connection: Connection, authkey: bytes) -> None: ...
+def answer_challenge(connection: Connection[Any, Any], authkey: bytes) -> None: ...
 def wait(
-    object_list: Iterable[Connection | socket.socket | int], timeout: float | None = None
-) -> list[Connection | socket.socket | int]: ...
-def Client(address: _Address, family: str | None = None, authkey: bytes | None = None) -> Connection: ...
+    object_list: Iterable[Connection[_T1, _T2] | socket.socket | int], timeout: float | None = None
+) -> list[Connection[_T1, _T2] | socket.socket | int]: ...
+def Client(address: _Address, family: str | None = None, authkey: bytes | None = None) -> Connection[Incomplete, Incomplete]: ...
 
 # N.B. Keep this in sync with multiprocessing.context.BaseContext.Pipe.
 # _ConnectionBase is the common base class of Connection and PipeConnection
 # and can be used in cross-platform code.
 if sys.platform != "win32":
-    def Pipe(duplex: bool = True) -> tuple[Connection, Connection]: ...
+    def Pipe(duplex: bool = True) -> tuple[Connection[_T1, _T2], Connection[_T2, _T1]]: ...
 
 else:
-    def Pipe(duplex: bool = True) -> tuple[PipeConnection, PipeConnection]: ...
+    def Pipe(duplex: bool = True) -> tuple[PipeConnection[_T1, _T2], PipeConnection[_T2, _T1]]: ...
```

Comment:
As seen in the primer hit (and the tests), this means that the return types of `Pipe()` need to be annotated now, which is a bit cumbersome.

Reply:
I'm not sure it's actually valid to use a TypeVar like this, where it's entirely scoped within a return annotation. IIRC type checkers special-case Callable so that you can do things like `def x() -> Callable[[T], T]`, which comes up a lot. But in general, I don't think there's any supported way of spelling what you're trying to express here, unfortunately :(

Reply:
As shown in the tests. Mypy doesn't support this to restrict the generics between each other (but it also doesn't fail). Pyright supports it, but idk if intentionally or not.
I don't see anything about it either in https://peps.python.org/pep-0695/ or https://peps.python.org/pep-0696/
I was hoping `def Pipe[_T1, _T2](duplex: bool = True) -> tuple[PipeConnection[_T1, _T2], PipeConnection[_T2, _T1]]: ...` would eventually be supported syntax.

Reply:
I'd rather not do this for now. It's an underspecified area of the spec, and we can see from the primer hits that mypy can't currently solve the TypeVars if we do things like this, causing false positives in user code

Reply:
I've moved the expectation of inverting the generic parameters, from the type definition, to a comment.